### PR TITLE
gNOI Target Verify (#114)

### LINKED ISSUE #114: Implement OS Verify RPC for mock target

### PR DESCRIPTION
**Pull request by Jared-Mullin**
Related Issue: Closes #114
Implementation of Verify service
Added flags to gnoi_target.go 
Updated README with new flags
Added new activationFailMessage to os.Server struct, which the target can be initialised with and cleared upon successful OS activation

### CODE REVIEW

File: gnoi/os/server_test.go [hunk 1]
```diff
@@ -57,8 +47,45 @@ func TestActivate(t *testing.T) {
 	}
 	for _, test := range tests {
 		got, _ := server.Activate(context.Background(), test.request)
-		if diff := pretty.Compare(test.want.Response, got.Response); diff != "" {
+		diff := pretty.Compare(test.want.Response, got.Response)
+		if diff != "" {
```
Comment (loriscn12) [inline gnoi/os/server_test.go:51]: was this reverted by mistake?

File: gnoi/os/server_test.go [hunk 2]
```diff
@@ -57,8 +47,45 @@ func TestActivate(t *testing.T) {
...
+				Version:               "2.0.1b",
+				ActivationFailMessage: "This is a test activation_fail_message",
+			},
+		},
+	}
+	for _, test := range tests {
+		server := NewServer(test.settings)
+		got, _ := server.Verify(context.Background(), &pb.VerifyRequest{})
+		diff := pretty.Compare(test.want, got)
+		if diff != "" {
```
Comment (loriscn12) [inline gnoi/os/server_test.go:87]: you can remove one line.

File: gnoi/os/server_test.go (hunk 1 above)
Comment (Jared-Mullin) [reply gnoi/os/server_test.go]: Yes, not exactly sure how

**Comment by samribeiro** [comment]
os.Activation will only fail if the version requested does not exist.

**Comment by Jared-Mullin** [comment]
Sure, I understand! I'll make those changes now

File: gnoi/os/server.go [hunk 3]
```diff
@@ -25,15 +22,19 @@ import (
 // Server is an OS Management service.
 type Server struct {
 	pb.OSServer
-	manager *Manager
+	manager               *Manager
+	activationFailMessage string
```
Comment (samribeiro) [inline gnoi/os/server.go:26]: probably better to have this in manager, it avoids spliting the logic between the files.
server is more like a grpc abstraction
manager is more like the os abstraction

File: gnoi/os/manager.go [hunk 4]
```diff
@@ -22,6 +19,7 @@ import (
 // Manager for storing data on OS's.
 type Manager struct {
 	osMap          map[string]bool
+	failMsgs       map[string]string
```
Comment (samribeiro) [inline gnoi/os/manager.go:22]: also:
activationFailMessage string

Comment (samribeiro) [reply gnoi/os/manager.go:22]: then below:

```
// SetRunning sets the running OS to the version specified.
func (m *Manager) SetRunning(version string) error {
        if m.activationFailMessage = s.manager.failMsgs[request.Version]; m.activationFailMessage != "" {
		return nil
        }
	if _, ok := m.osMap[version]; ok {
		m.runningVersion = version
		return nil
	}
	return fmt.Errorf("NON_EXISTENT_VERSION")
}
```

Comment (samribeiro) [reply gnoi/os/manager.go:22]: Also you can add the failMessage as a new parameter in the Install function.

File: gnoi/os/server.go (hunk 3 above)
Comment (Jared-Mullin) [reply gnoi/os/server.go:26]: That makes sense! Will do

File: gnoi/os/manager.go (hunk 4 above)
Comment (Jared-Mullin) [reply gnoi/os/manager.go:22]: Added!

File: utils/mockos/mockos.go [hunk 5]
```diff
@@ -19,7 +19,7 @@ import (
 	"errors"
 	"os"
 
-	humanize "github.com/dustin/go-humanize"
```
[the remark this answers is not in the document]
Comment (Jared-Mullin) [reply utils/mockos/mockos.go:22]: :D

File: utils/mockos/pb/mockos.proto [hunk 6]
```diff
@@ -20,5 +20,6 @@ message MockOS {
     string cookie = 2; // Unique identifier of OS.
     bytes padding = 3; // Random bytes to fill size.
     bytes hash = 4; 
-    bool unsupported = 5; // If true then the OS package is valid but unsupported by the mock target.
+    bool incompatible = 5; // If true then the OS package is valid but incompatible with the mock target.
+    string activation_fail_message = 6;
```
[the remark this answers is not in the document]
Comment (Jared-Mullin) [reply utils/mockos/pb/mockos.proto:24]: Thanks! Added

**Pull request merged**